### ISSUE TITLE
Stripping annotations also removes the portable PDB checksum

### ISSUE THREAD

**Issue by rdeago** [tenacom/ReSharper.ExportAnnotations#13]
## Description

**Version:** 1.1.0

When `ExportAnnotations` modifies an assembly, the portable PDB checksum disappears. This causes NuGet to reject uploaded symbol package files, as they fail the checksum test due to the fact that there is no checksum to begin with.

This is caused by jbevain/cecil#610 which is also affecting Fody and other projects.

Workarounds proposed so far include embedding debug information in assemblies, and including PDB files in .nupkg packages. None of these applies here, of course, because we cannot make such decisions on the behalf of users.

## Configuration

Reproducible by referencing the `ReSharper.ExportAnnotations` package and setting the `StripJetBrainsAnnotations` property to `true`.

## Regression?

Not a regression. It's always been this way, but it apparently never affected anyone until I tried to distribute symbol packages for the compiled tasks in [Buildvana.Sdk](https://github.com/Buildvana/Buildvana.Sdk).

## Other information

A possible workaround could be: instead of stripping annotations for the assembly, delete it and re-run the `Build` target without the `JETBRAINS_ANNOTATIONS` constant. Some quick notes:

- Building twice has an obvious negative impact on performance.
- We should be able to call the Build target again and have MSBuild not skip it just by setting a different set of properties (which we have to do anyway). That's the way multi-targeting works.
- This would also close #2 and any other possible problems with assembly modification.

**Comment by rdeago** [comment]
## How do we "re-run" a target in MSBuild?

**TL;DR:** working on it. No promises.

## Hypothesis 1: use the MSBuild task

### Procedure

After running the `ExportAnnotations` tool:

1. Delete the compiled assembly from the intermediate output directory.
2. Remove `JETBRAINS_ANNOTATIONS` from the `DefineConstants` property.
3. Use the `MSBuild` task to invoke `CoreCompile` again, so it produces an assembly without annotations.

### Research notes

This goes down the drain before we even try. Quoting the [documentation for the MSBuild task](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-task) (emphasis mine):

> Unlike using the Exec task to start MSBuild.exe, this task uses the same MSBuild process to build the child projects. **The list of already-built targets that can be skipped is shared between the parent and child builds.**

Calling a target more than once in the same execution is something MSBuild was built from the beginning to prevent.

But even if we could, there's one more problem with re-invoking `CoreCompile`: it [erases one of its inputs](https://github.com/dotnet/roslyn/blob/c64d1e3a7b3e2b6d5a71ca4ceb5848bcd79a0970/src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets#L154-L156), so a second invocation would either fail, or produce a different assembly than the one we want. This goes to suggest that, as a general rule, we should only try to re-invoke "public-facing" targets such as `Build`.

### Conclusion

Hypothesis dismissed as unfeasible.

## Hypothesis 2: run another MSBuild instance

### Procedure

In a target set to run immediately after `Build` if `StripJetBrainsAnnotations` is `true`:

1. Retrieve the original MSBuild command line.
2. Modify the command line as follows:
    - invoke the `Build`  target instead of whichever target(s) was (were) originally invoked;
    - add a property that lets us know that this is the second invocation of `Build`.
3. Delete the compiled asembly from the output directory (from the intermediate directory as well, just to be sure).
4. Run another MSBuild instance to rebuild the assembly without ReSharper annotations.
5. Go on with whatever MSBuild was doing.

If this is the second invocation, as indicated by the property we set in step 2 above:

- Don't add `JETBRAINS_ANNOTATIONS` to `DefineConstants`.
- Set `GeneratePackageOnBuild` to `false`, as there's no point in invoking `Pack` twice.
- Do not run the `ExportAnnotations` tool.

### Research notes

[This StackOverflow answer](https://stackoverflow.com/a/52190417) explains how to retrieve the original command line. It requires an inline (or, why not, compiled) task so while we're there we can also parse and modify the command line in C#, then use a `Process` to run MSBuild.

Regarding command line parameters, we'll have to do more than just changing the target and adding a property. A quick glance at [MSBuild command-line reference](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-command-line-reference) reveals the following:

- We should remove any reference to logging, profile evaluation, and graph building.
- We should try to respect `maxCpuCount`, although it is not clear to me just how - maybe just set it to 1 for safety.
- What about response files? Or worse, [autoresponse files](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-response-files)? Looks like we should parse them as well, then build a command line that does not involve them.

### Conclusion

This looks promising, although the command line parsing part is probably going to be a major PITA.

Let's keep this door open for now.

---

I had two more hypotheses in mind, both involving duplication of Roslyn's `CoreCompile` targets. As we've seen, though, messing with `CoreCompile` is not the way to go.

**Comment by rdeago** [comment]
Fixed by #14. Closing.
